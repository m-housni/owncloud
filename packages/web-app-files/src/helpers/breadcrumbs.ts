import { eventBus } from 'web-pkg/src/services/eventBus'
import { Location } from 'vue-router'
import omit from 'lodash-es/omit'

export interface BreadcrumbItem {
  text: string
  to?: Location
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentRoute: Location,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentRoute.path)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: {
          path: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/'),
          query: omit(currentRoute.query, 'fileId', 'page') // TODO: we need the correct fileId in the query. until we have that we must omit it because otherwise we would correct the path to the one of the (wrong) fileId.
        }
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => eventBus.publish('app.files.list.load')
    }
  ]
}
