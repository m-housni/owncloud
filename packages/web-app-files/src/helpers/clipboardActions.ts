export abstract class ClipboardActions {
  static readonly Cut = 'cut'
  static readonly Copy = 'copy'
}
