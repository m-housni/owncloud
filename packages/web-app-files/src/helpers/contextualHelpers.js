// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}
export const empty = {
  text: '',
  list: ['', '', '']
}
export const shareInviteCollaboratorHelp = {
  title: $gettext('Invite specific people'),
  text: $gettext('Enter a name or group to share this item.'),
  list: [
    { text: $gettext('Subfolders'), headline: true },
    {
      text: $gettext(
        'If you share a folder, all of its contents and subfolders will be shared as well.'
      )
    },
    { text: $gettext('Notification'), headline: true },
    {
      text: $gettext('Invited people will be notified via email or in-app notification.')
    },
    { text: $gettext('Incognito'), headline: true },
    {
      text: $gettext('Invited people can not see who else has access..')
    },
    { text: $gettext('“via folder”'), headline: true },
    {
      text: $gettext(
        'The “via folder” is shown next to a share, if access has already been given via a parent folder. Click on the “via folder” to edit the share on its parent folder.'
      )
    }
  ],
  readMoreLink: 'https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing'
}
export const shareInviteCollaboratorHelpCern = {
  list: [
    { text: $gettext('Search for service or secondary Account'), headline: true },
    {
      text: $gettext(
        'To search for service or secondary accounts prefix the username with "a:" (like "a:doe") and for guest accounts prefix the username with "l:" (like "l:doe").'
      )
    }
  ]
}
export const shareSpaceAddMemberHelp = {
  title: $gettext('Add members to this Space'),
  text: $gettext('Enter a name to add people or groups as members to this Space.'),
  list: [
    { text: $gettext('What members can do'), headline: true },
    {
      text: $gettext(
        'Members can see who else has access to this space and can access all files in this space. Read or write permissions can be set by the member’s role such as “Viewer” or “Editor”.'
      )
    },
    { text: $gettext('What Space managers can do'), headline: true },
    {
      text: $gettext(
        'Members with the Manager role can edit all properties and content of a Space, such as adding or removing members, sharing subfolders with non-members, or creating links to share.'
      )
    }
  ],
  readMoreLink: 'https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing'
}
export const shareViaLinkHelp = {
  title: $gettext('Choose how access is granted'),
  list: [
    {
      text: $gettext('Only invited people can access'),
      headline: true
    },
    {
      text: $gettext(
        'Account and login is required. Only people from the list "Invited people" can access.'
      )
    },
    {
      text: $gettext('Everyone with the link'),
      headline: true
    },
    {
      text: $gettext(
        'No login required. Everyone with the link can access. If you share this link with people from the list "Invited people", they need to login so that their individual assigned permissions can take effect. If they are not logged-in, the permissions of the link take effect.'
      )
    },
    {
      text: $gettext('Quicklink'),
      headline: true
    },
    {
      text: $gettext(
        'The quicklink is the default link that is copied when you select "Get link” from the context menu.'
      )
    }
  ],
  readMoreLink: 'https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing'
}
export const shareViaIndirectLinkHelp = {
  title: $gettext('What are indirect links?'),
  text: $gettext('Indirect links are links giving access by a parent folder.'),
  list: [
    {
      text: $gettext('How to edit indirect links'),
      headline: true
    },
    {
      text: $gettext(
        'Indirect links can only be edited in their parent folder. Click on the folder icon below the link to navgate to the parent folder.'
      )
    }
  ],
  readMoreLink: 'https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing'
}
