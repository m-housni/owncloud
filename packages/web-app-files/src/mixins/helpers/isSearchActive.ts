export default {
  computed: {
    $_isSearchActive(): boolean {
      return !!document.getElementById('files-global-search-options')
    }
  }
}
