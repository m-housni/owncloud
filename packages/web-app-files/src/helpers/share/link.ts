import { DateTime } from 'luxon'
import { Share } from 'web-client/src/helpers/share'
import { Store } from 'vuex'
import { clientService } from 'web-pkg/src/services'
import copyToClipboard from 'copy-to-clipboard'

interface CreateQuicklink {
  store: Store<any>
  storageId?: any
  resource: any
  password?: string
  $gettext: (string) => string
}

export const createQuicklink = async (args: CreateQuicklink): Promise<Share> => {
  const { resource, store, password, $gettext } = args
  const params: { [key: string]: unknown } = {
    name: $gettext('Quicklink'),
    permissions: 1,
    quicklink: true
  }

  if (password) {
    params.password = password
  }

  const expirationDate = store.state.user.capabilities.files_sharing.public.expire_date

  if (expirationDate.enforced) {
    params.expireDate = DateTime.now()
      .plus({ days: parseInt(expirationDate.days, 10) })
      .endOf('day')
      .toISO()
  }

  // needs check for enforced password for default role (viewer?)
  // and concept to what happens if it is enforced

  params.spaceRef = resource.fileId || resource.id

  const link = await store.dispatch('Files/addLink', {
    path: resource.path,
    client: clientService.owncloudSdk,
    params,
    storageId: resource.fileId || resource.id
  })

  copyToClipboard(link.url)

  await store.dispatch('showMessage', {
    title: $gettext('The quicklink has been copied to your clipboard.')
  })

  return link
}
