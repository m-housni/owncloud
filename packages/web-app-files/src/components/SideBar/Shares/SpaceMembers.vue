<template>
  <div id="oc-files-sharing-sidebar" class="oc-position-relative">
    <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
      <div class="oc-flex oc-py-s">
        <h3 v-translate class="oc-text-bold oc-text-medium oc-m-rm">Members</h3>
        <oc-contextual-helper v-if="helpersEnabled" class="oc-pl-xs" v-bind="spaceAddMemberHelp" />
      </div>
      <div class="oc-flex">
        <div v-if="isFilterOpen" class="oc-flex">
          <oc-text-input
            v-model="filterTerm"
            class="oc-text-truncate space-members-filter oc-mr-s"
            label=""
            :placeholder="$gettext('Filter members')"
          />
          <oc-button
            v-oc-tooltip="$gettext('Close filter')"
            :aria-label="$gettext('Close filter')"
            appearance="raw"
            @click="toggleFilter"
          >
            <oc-icon name="close" fill-type="line" size="small" />
          </oc-button>
        </div>
        <oc-button
          v-else
          v-oc-tooltip="$gettext('Filter members')"
          :aria-label="$gettext('Filter members')"
          appearance="raw"
          @click="toggleFilter"
        >
          <oc-icon name="search" fill-type="line" size="small" />
        </oc-button>
      </div>
    </div>
    <invite-collaborator-form
      v-if="currentUserCanShare"
      key="new-collaborator"
      :save-button-label="$gettext('Add')"
      :invite-label="$gettext('Add members')"
      class="oc-my-s"
    />
    <template v-if="hasCollaborators">
      <ul
        id="files-collaborators-list"
        ref="collaboratorList"
        class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm"
        :aria-label="$gettext('Space members')"
      >
        <li v-for="collaborator in filteredSpaceMembers" :key="collaborator.key">
          <collaborator-list-item
            :share="collaborator"
            :modifiable="isModifiable(collaborator)"
            @onDelete="$_ocCollaborators_deleteShare_trigger(collaborator)"
          />
        </li>
      </ul>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapState } from 'vuex'
import CollaboratorListItem from './Collaborators/ListItem.vue'
import InviteCollaboratorForm from './Collaborators/InviteCollaborator/InviteCollaboratorForm.vue'
import { spaceRoleManager } from 'web-client/src/helpers/share'
import { createLocationSpaces, isLocationSpacesActive } from '../../../router'
import { defineComponent, PropType } from 'vue'
import { shareSpaceAddMemberHelp } from '../../../helpers/contextualHelpers'
import { SpaceResource } from 'web-client/src/helpers'
import { useGraphClient } from 'web-pkg/src/composables'
import Fuse from 'fuse.js'
import Mark from 'mark.js'

export default defineComponent({
  name: 'SpaceMembers',
  components: {
    CollaboratorListItem,
    InviteCollaboratorForm
  },
  props: {
    space: {
      type: Object as PropType<SpaceResource>,
      required: false,
      default: null
    }
  },
  setup() {
    return {
      ...useGraphClient()
    }
  },
  data: () => {
    return {
      filterTerm: '',
      isFilterOpen: false,
      markInstance: null
    }
  },
  computed: {
    ...mapGetters(['configuration']),
    ...mapGetters('runtime/spaces', ['spaceMembers']),
    ...mapState(['user']),

    filteredSpaceMembers() {
      return this.filter(this.spaceMembers, this.filterTerm)
    },
    helpersEnabled() {
      return this.configuration?.options?.contextHelpers
    },
    spaceAddMemberHelp() {
      return shareSpaceAddMemberHelp
    },
    hasCollaborators() {
      return this.spaceMembers.length > 0
    },
    currentUserCanShare() {
      return this.currentUserIsManager
    },
    currentUserIsManager() {
      const currentUserCollaborator = this.spaceMembers.find(
        (collaborator) => collaborator.collaborator.name === this.user.id
      )

      return currentUserCollaborator?.role?.name === spaceRoleManager.name
    }
  },
  watch: {
    isFilterOpen() {
      this.filterTerm = ''
    },
    filterTerm() {
      this.$nextTick(() => {
        if (this.$refs.collaboratorList) {
          this.markInstance = new Mark(this.$refs.collaboratorList)
          this.markInstance.unmark()
          this.markInstance.mark(this.filterTerm, {
            element: 'span',
            className: 'highlight-mark'
          })
        }
      })
    }
  },
  methods: {
    ...mapActions('runtime/spaces', ['deleteSpaceMember']),
    ...mapActions(['createModal', 'hideModal', 'showMessage']),

    filter(collection, term) {
      if (!(term || '').trim()) {
        return collection
      }
      const searchEngine = new Fuse(collection, {
        includeScore: true,
        useExtendedSearch: true,
        threshold: 0.3,
        keys: ['collaborator.displayName', 'collaborator.name']
      })

      return searchEngine.search(term).map((r) => r.item)
    },
    toggleFilter() {
      this.isFilterOpen = !this.isFilterOpen
    },
    isModifiable(share) {
      if (!this.currentUserIsManager) {
        return false
      }

      if (share.role.name !== spaceRoleManager.name) {
        return true
      }

      // forbid to remove last manager of a space
      const managers = this.spaceMembers.filter(
        (collaborator) => collaborator.role.name === spaceRoleManager.name
      )
      return managers.length > 1
    },

    $_ocCollaborators_deleteShare_trigger(share) {
      const modal = {
        variation: 'danger',
        icon: 'alarm-warning',
        title: this.$gettext('Remove share'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.$gettext('Remove'),
        message: this.$gettext('Are you sure you want to remove this share?'),
        hasInput: false,
        onCancel: this.hideModal,
        onConfirm: () => this.$_ocCollaborators_deleteShare(share)
      }

      this.createModal(modal)
    },

    async $_ocCollaborators_deleteShare(share) {
      try {
        const currentUserRemoved = share.collaborator.name === this.user.id
        await this.deleteSpaceMember({
          client: this.$client,
          graphClient: this.graphClient,
          share: share,
          reloadSpace: !currentUserRemoved
        })
        this.showMessage({
          title: this.$gettext('Share was removed successfully')
        })

        if (currentUserRemoved) {
          if (isLocationSpacesActive(this.$router, 'files-spaces-projects')) {
            return this.$router.go()
          }
          return this.$router.push(createLocationSpaces('files-spaces-projects'))
        }
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettext('Failed to remove share'),
          status: 'danger'
        })
      } finally {
        this.hideModal()
      }
    }
  }
})
</script>

<style>
.highlight-mark {
  font-weight: 600;
}
.space-members-filter {
  max-width: 160px;
}
.avatars-wrapper {
  height: 40px;
}
</style>
