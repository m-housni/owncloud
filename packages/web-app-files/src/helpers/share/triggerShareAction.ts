import { aggregateResourceShares } from '../resources'
import { ShareStatus } from 'web-client/src/helpers/share/status'

export async function triggerShareAction(resource, status, hasReSharing, hasShareJail, $client) {
  const method = _getRequestMethod(status)
  if (!method) {
    throw new Error('invalid new share status')
  }

  // exec share action
  let response = await $client.requests.ocs({
    service: 'apps/files_sharing',
    action: `api/v1/shares/pending/${resource.share.id}`,
    method
  })

  // exit on failure
  if (response.status !== 200) {
    throw new Error(response.statusText)
  }

  // get updated share from response and transform & return it
  if (parseInt(response.headers.get('content-length')) > 0) {
    response = await response.json()
    if (response.ocs.data.length > 0) {
      const share = response.ocs.data[0]
      return aggregateResourceShares([share], true, hasReSharing, hasShareJail)[0]
    }
  }

  return null
}

function _getRequestMethod(status) {
  switch (status) {
    case ShareStatus.accepted:
      return 'POST'
    case ShareStatus.declined:
      return 'DELETE'
    default:
      return null
  }
}
