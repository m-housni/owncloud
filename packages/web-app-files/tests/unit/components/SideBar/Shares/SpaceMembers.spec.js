import SpaceMembers from 'web-app-files/src/components/SideBar/Shares/SpaceMembers.vue'
import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import GetTextPlugin from 'vue-gettext'
import Vuex from 'vuex'
import DesignSystem from '@ownclouders/design-system'
import Users from '@/__fixtures__/users'
import { spaceRoleManager, spaceRoleViewer, ShareTypes } from 'web-client/src/helpers/share'

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const user = Users.alice
const spaceMock = {
  type: 'space',
  name: ' space',
  id: '1',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  spaceQuota: {
    used: 100,
    total: 1000
  }
}

const spaceMembers = [
  {
    id: '1',
    shareType: ShareTypes.space.value,
    collaborator: {
      name: user.id,
      displayName: user.displayname
    },
    role: {
      name: spaceRoleManager.name
    }
  },
  {
    id: '2',
    shareType: ShareTypes.space.value,
    collaborator: {
      onPremisesSamAccountName: 'Einstein',
      displayName: 'einstein'
    },
    role: {
      name: spaceRoleViewer.name
    }
  },
  {
    id: '3',
    shareType: ShareTypes.space.value,
    collaborator: {
      onPremisesSamAccountName: 'Marie',
      displayName: 'marie'
    },
    role: {
      name: spaceRoleManager.name
    }
  }
]

describe('SpaceMembers', () => {
  it('renders loading spinner while loading', () => {
    const wrapper = getShallowMountedWrapper(
      {
        user
      },
      true
    )
    expect(wrapper).toMatchSnapshot()
  })
  describe('if currentUser can share', () => {
    it('initially renders add people dialog', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        spaceMembers: [spaceMembers[0]]
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if currentUser can not share', () => {
    it('other shares are listed, but creating/editing shares is not possible', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        spaceMembers: [spaceMembers[1]]
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if currentUser is manager', () => {
    it('allows role edit of the current user if another user is manager', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        spaceMembers
      })
      expect(wrapper).toMatchSnapshot()
    })
    it('does not allow role edit of the current user if they are the only manager', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        spaceMembers: [spaceMembers[0], spaceMembers[1]]
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if currentUser can delete', () => {
    it('reacts on delete events by collaborator list items', async () => {
      const wrapper = getMountedWrapper({
        user,
        spaceMembers
      })

      const spyOnCollaboratorDeleteTrigger = jest.spyOn(
        wrapper.vm,
        '$_ocCollaborators_deleteShare_trigger'
      )
      wrapper
        .find(`div[data-testid="collaborator-user-item-${spaceMembers[0].collaborator.name}"]`)
        .vm.$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(spyOnCollaboratorDeleteTrigger).toHaveBeenCalledTimes(1)
    })
  })
})

const storeOptions = (data, isInLoadingState) => {
  const { user, spaceMembers = [] } = data

  return {
    actions: {
      createModal: jest.fn(),
      hideModal: jest.fn(),
      showMessage: jest.fn()
    },
    state: {
      user
    },
    modules: {
      Files: {
        namespaced: true,
        getters: {
          highlightedFile: () => spaceMock,
          sharesTreeLoading: () => false
        },
        actions: {
          deleteShare: jest.fn()
        },
        mutations: {
          SET_HIGHLIGHTED_FILE(state, file) {
            state.highlightedFile = spaceMock
          }
        }
      },
      runtime: {
        namespaced: true,
        modules: {
          spaces: {
            namespaced: true,
            getters: {
              spaceMembers: () => spaceMembers
            }
          }
        }
      }
    },
    getters: {
      configuration: jest.fn(() => ({
        server: 'http://example.com/',
        currentTheme: {
          general: {
            name: 'some-company'
          }
        }
      })),
      user: () => user,
      capabilities: () => {
        return {
          files_sharing: {
            user: {
              expire_date: {
                enabled: true,
                days: 10
              }
            },
            group: {
              expire_date: {
                enabled: true,
                days: 10
              }
            }
          }
        }
      }
    }
  }
}

function getShallowMountedWrapper(data, loading = false) {
  return shallowMount(SpaceMembers, {
    localVue,
    store: createStore(data, loading),
    stubs: {
      'oc-button': true,
      'oc-icon': true,
      'oc-spinner': true
    },
    provide: {
      displayedItem: {
        value: spaceMock
      }
    }
  })
}

function getMountedWrapper(data, loading = false) {
  return mount(SpaceMembers, {
    localVue,
    store: createStore(data, loading),
    stubs: {
      'oc-button': true,
      'oc-icon': true,
      'oc-spinner': true,
      'avatar-image': true,
      'role-dropdown': true,
      'edit-dropdown': true
    },
    mocks: {
      $router: {
        go: jest.fn(),
        push: jest.fn(),
        currentRoute: {
          name: 'some-route',
          query: { page: 1 }
        },
        resolve: (r) => ({ href: r.name })
      }
    },
    provide: {
      displayedItem: {
        value: spaceMock
      }
    }
  })
}

function createStore(data, loading) {
  return new Vuex.Store(storeOptions(data, loading))
}
