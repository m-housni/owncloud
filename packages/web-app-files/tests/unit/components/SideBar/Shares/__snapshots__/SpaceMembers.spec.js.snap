// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceMembers if currentUser can not share other shares are listed, but creating/editing shares is not possible 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
    <div class="oc-flex oc-py-s">
      <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Members" data-current-language="en_US">Members</h3>
      <!---->
    </div>
    <div class="oc-flex">
      <oc-button-stub type="button" size="medium" arialabel="Filter members" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
        <oc-icon-stub name="search" filltype="line" accessiblelabel="" type="span" size="small" variation="passive" color=""></oc-icon-stub>
      </oc-button-stub>
    </div>
  </div>
  <!---->
  <ul id="files-collaborators-list" aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
</div>
`;

exports[`SpaceMembers if currentUser can share initially renders add people dialog 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
    <div class="oc-flex oc-py-s">
      <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Members" data-current-language="en_US">Members</h3>
      <!---->
    </div>
    <div class="oc-flex">
      <oc-button-stub type="button" size="medium" arialabel="Filter members" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
        <oc-icon-stub name="search" filltype="line" accessiblelabel="" type="span" size="small" variation="passive" color=""></oc-icon-stub>
      </oc-button-stub>
    </div>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Add" invitelabel="Add members" class="oc-my-s"></invite-collaborator-form-stub>
  <ul id="files-collaborators-list" aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
</div>
`;

exports[`SpaceMembers if currentUser is manager allows role edit of the current user if another user is manager 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
    <div class="oc-flex oc-py-s">
      <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Members" data-current-language="en_US">Members</h3>
      <!---->
    </div>
    <div class="oc-flex">
      <oc-button-stub type="button" size="medium" arialabel="Filter members" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
        <oc-icon-stub name="search" filltype="line" accessiblelabel="" type="span" size="small" variation="passive" color=""></oc-icon-stub>
      </oc-button-stub>
    </div>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Add" invitelabel="Add members" class="oc-my-s"></invite-collaborator-form-stub>
  <ul id="files-collaborators-list" aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
  </ul>
</div>
`;

exports[`SpaceMembers if currentUser is manager does not allow role edit of the current user if they are the only manager 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
    <div class="oc-flex oc-py-s">
      <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Members" data-current-language="en_US">Members</h3>
      <!---->
    </div>
    <div class="oc-flex">
      <oc-button-stub type="button" size="medium" arialabel="Filter members" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
        <oc-icon-stub name="search" filltype="line" accessiblelabel="" type="span" size="small" variation="passive" color=""></oc-icon-stub>
      </oc-button-stub>
    </div>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Add" invitelabel="Add members" class="oc-my-s"></invite-collaborator-form-stub>
  <ul id="files-collaborators-list" aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
  </ul>
</div>
`;

exports[`SpaceMembers renders loading spinner while loading 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
    <div class="oc-flex oc-py-s">
      <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Members" data-current-language="en_US">Members</h3>
      <!---->
    </div>
    <div class="oc-flex">
      <oc-button-stub type="button" size="medium" arialabel="Filter members" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
        <oc-icon-stub name="search" filltype="line" accessiblelabel="" type="span" size="small" variation="passive" color=""></oc-icon-stub>
      </oc-button-stub>
    </div>
  </div>
  <!---->
  <!---->
</div>
`;
